mlflow/tracking/client.py

```diff
@@ -1010,6 +1010,62 @@ def log_dict(self, run_id, dictionary, artifact_file):
                 else:
                     json.dump(dictionary, f, indent=2)
 
+    def log_figure(self, run_id, figure, artifact_file):
+        """
+        Log a figure as an artifact. The following figure objects are supported:
+
+        - `matplotlib.figure.Figure`_
+        - `plotly.graph_objects.Figure`_
+
+        .. _matplotlib.figure.Figure:
+            https://matplotlib.org/api/_as_gen/matplotlib.figure.Figure.html
+
+        .. _plotly.graph_objects.Figure:
+            https://plotly.com/python-api-reference/generated/plotly.graph_objects.Figure.html
+
+        :param run_id: String ID of the run.
+        :param figure: Figure to log.
+        :param artifact_file: The run-relative artifact file path in posixpath format to which
+                              the figure is saved (e.g. "dir/file.png").
+
+        .. code-block:: python
+            :caption: Matplotlib Example
+
+            import mlflow
+            import matplotlib.pyplot as plt
+
+            fig, ax = plt.subplots()
+            ax.plot([0, 1], [2, 3])
+
+            run = client.create_run(experiment_id="0")
+            client.log_figure(run.info.run_id, fig, "figure.png")
+
+        .. code-block:: python
+            :caption: Plotly Example
+
+            import mlflow
+            from plotly import graph_objects as go
+
+            fig = go.Figure(go.Scatter(x=[0, 1], y=[2, 3]))
+
+            run = client.create_run(experiment_id="0")
+            client.log_figure(run.info.run_id, fig, "figure.html")
+        """
+        try:
+            import plotly
+            from matplotlib import pyplot as plt
+            from plotly import graph_objects as go
```

Comment:
Instead of requiring  `plotly` to log a matplotlib figure and `matplotlib` to log a `plotly` figure (because we import both here), could we just assume e.g. a matplotlib figure and wrap the `savefig` call in a `try ... except AttributeError`, falling back to calling `plotly.offline.plot`? 

Or maybe alternatively we could perform a check like `if "plotly" in sys.modules` and only check whether the object is a `plotly` figure if `plotly` has already been imported, etc

Reply:
The try-catch approach might result in a confusing error message:

```python
fig = ""  # user thinks `fig` is a matplotlib figure but actually is not.

try:
    fig.savefig("foo.png")
except AttributeError:
    plotly.offline.plot(fig, filename="foo.html", include_plotlyjs="cdn", auto_open=False)
```

This code results in:

```
Traceback (most recent call last):
  File "a.py", line 5, in <module>
    fig.savefig("foo.png")
AttributeError: 'str' object has no attribute 'savefig'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "a.py", line 7, in <module>
    plotly.offline.plot(fig, filename="foo.html", include_plotlyjs="cdn", auto_open=False)
  File "/Users/harutakakawamura/.pyenv/versions/miniconda3-4.7.12/envs/mlflow-dev-env/lib/python3.6/site-packages/plotly/offline/offline.py", line 573, in plot
    figure = tools.return_figure_from_figure_or_data(figure_or_data, validate)
  File "/Users/harutakakawamura/.pyenv/versions/miniconda3-4.7.12/envs/mlflow-dev-env/lib/python3.6/site-packages/plotly/tools.py", line 545, in return_figure_from_figure_or_data
    "The `figure_or_data` positional "
_plotly_utils.exceptions.PlotlyError: The `figure_or_data` positional argument must be `dict`-like, `list`-like, or an instance of plotly.graph_objs.Figure
```

Reply:
This is what neptune.ai does:

https://neptune-contrib.readthedocs.io/_modules/neptunecontrib/api/chart.html#log_chart